internal/controller/dnsrecord_controller.go

```diff
@@ -152,22 +152,43 @@ func (r *DNSRecordReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		return ctrl.Result{}, err
 	}
 
-	return r.updateStatus(ctx, previous, dnsRecord, requeueAfter)
+	return r.updateStatus(ctx, previous, dnsRecord, requeueAfter, nil)
 }
 
-func (r *DNSRecordReconciler) updateStatus(ctx context.Context, previous, current *v1alpha1.DNSRecord, requeueAfter time.Duration) (reconcile.Result, error) {
-	current.Status.QueuedFor = metav1.NewTime(reconcileStart.Add(requeueAfter))
+func (r *DNSRecordReconciler) updateStatus(ctx context.Context, previous, current *v1alpha1.DNSRecord, requeueAfter time.Duration, specErr error) (reconcile.Result, error) {
+	r.calculateStatus(current, requeueAfter, specErr)
 
 	if !equality.Semantic.DeepEqual(previous.Status, current.Status) {
-		updateError := r.Status().Update(ctx, current)
-		if apierrors.IsConflict(updateError) {
-			return ctrl.Result{Requeue: true}, nil
+		if updateError := r.Status().Update(ctx, current); updateError != nil {
+			if apierrors.IsConflict(updateError) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+			return ctrl.Result{}, updateError
 		}
-		return ctrl.Result{RequeueAfter: requeueAfter}, updateError
 	}
+
+	if specErr != nil {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
 	return ctrl.Result{RequeueAfter: requeueAfter}, nil
 }
 
+// calculateStatus calculate and apply the status of the dnsRecord based on what has happened during the reconcile.
+// ToDo This should be the only thing updating the status of the dnsRecord.
+func (r *DNSRecordReconciler) calculateStatus(dnsRecord *v1alpha1.DNSRecord, requeueAfter time.Duration, specErr error) {
+	if specErr != nil {
+		//Set these values to whatever they should be in an error state, they should not change again until we leave the error state.
+		dnsRecord.Status.WriteCounter = -1
```

Comment:
I'm not  sure we should reset the writeCounter, I'd think we leave that unchanged.

Reply:
What's the value in leaving it? Is it not meaningless in this scenario? We can possible leave it, but we need to ensure wherever it is getting changed doesn't change it on error or it potentially will keep updating which this what this is trying to avoid.

Reply:
yes, it should be unchanged in an error. 

The scenario I'm imagining is:
1) A controller is having to write it's changes a lot, incrementing this counter to make that visible
2) The controller gets ratelimited by the DNS provider (or network outage or whatever) temporarily.
3) With the above change, this error resets the counter to zero, but in reality the problem causing the multiple writes isn't resolved.

Reply:
OK, i can see that. We will need the changes i made in `applyChanges` in that case as currently it bumps this before even doing anything.

Reply:
updated